panic: runtime error: invalid memory address or nil pointer dereference

Can not start prometheus

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x28 pc=0x720455]

goroutine 262 [running]:
panic(0xd864c0, 0xc8200100d0)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/prometheus/prometheus/vendor/golang.org/x/net/context.propagateCancel(0x0, 0x0, 0x7fc937f05660, 0xc822fb0680)
	/go/src/github.com/prometheus/prometheus/vendor/golang.org/x/net/context/pre_go17.go:80 +0x25
github.com/prometheus/prometheus/vendor/golang.org/x/net/context.WithCancel(0x0, 0x0, 0x0, 0x0, 0xc820ae79a8)
	/go/src/github.com/prometheus/prometheus/vendor/golang.org/x/net/context/pre_go17.go:66 +0x126
github.com/prometheus/prometheus/retrieval.newScrapeLoop(0x0, 0x0, 0x7fc937ec0998, 0xc822fb83f0, 0x7fc937ec0968, 0xc82085c340, 0x7fc937ec0968, 0xc82085c360, 0x0, 0x0)
	/go/src/github.com/prometheus/prometheus/retrieval/scrape.go:380 +0x124
github.com/prometheus/prometheus/retrieval.(*scrapePool).sync(0xc822fc4480, 0xc820a5a0a0, 0x1, 0x1)
	/go/src/github.com/prometheus/prometheus/retrieval/scrape.go:228 +0x36d
github.com/prometheus/prometheus/retrieval.(*targetSet).sync(0xc822fc0190)
	/go/src/github.com/prometheus/prometheus/retrieval/targetmanager.go:258 +0x239
github.com/prometheus/prometheus/retrieval.(*targetSet).runProviders(0xc822fc0190, 0x7fc937f05690, 0xc820a56840, 0xc822fb6960)
	/go/src/github.com/prometheus/prometheus/retrieval/targetmanager.go:334 +0x34b
github.com/prometheus/prometheus/retrieval.(*TargetManager).reload(0xc8202ba9a0)
	/go/src/github.com/prometheus/prometheus/retrieval/targetmanager.go:124 +0x25e
github.com/prometheus/prometheus/retrieval.(*TargetManager).Run(0xc8202ba9a0)
	/go/src/github.com/prometheus/prometheus/retrieval/targetmanager.go:79 +0x1cf
created by main.Main
	/go/src/github.com/prometheus/prometheus/cmd/prometheus/main.go:172 +0x171f
```

Also has this problem, make me crazy.

```
INFO[0000] Starting prometheus (version=0.19.2, branch=master, revision=05b918a)  source=main.go:73
INFO[0000] Build context (go=go1.6.2, user=root@e19bc1d10c15, date=20160607-16:25:02)  source=main.go:74
INFO[0000] Loading configuration file prometheus.yml     source=main.go:206
INFO[0000] Loading series map and head chunks...         source=storage.go:341
INFO[0024] 1867719 series loaded.                        source=storage.go:346
WARN[0024] No AlertManagers configured, not dispatching any alerts  source=notifier.go:174
INFO[0024] Starting target manager...                    source=targetmanager.go:74
INFO[0024] Listening on :9090                            source=web.go:241
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x28 pc=0x723e05]

goroutine 202 [running]:
panic(0xd88cc0, 0xc82000e0f0)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/prometheus/prometheus/vendor/golang.org/x/net/context.propagateCancel(0x0, 0x0, 0x7ff3f7bb05c0, 0xc944c1f440)
	/go/src/github.com/prometheus/prometheus/vendor/golang.org/x/net/context/pre_go17.go:80 +0x25
github.com/prometheus/prometheus/vendor/golang.org/x/net/context.WithCancel(0x0, 0x0, 0x0, 0x0, 0xc94217b9a8)
	/go/src/github.com/prometheus/prometheus/vendor/golang.org/x/net/context/pre_go17.go:66 +0x126
github.com/prometheus/prometheus/retrieval.newScrapeLoop(0x0, 0x0, 0x7ff3f7b81058, 0xc944c42880, 0x7ff3f7b81028, 0xc90c534580, 0x7ff3f7b81028, 0xc90c5345a0, 0x0, 0x0)
	/go/src/github.com/prometheus/prometheus/retrieval/scrape.go:380 +0x124
github.com/prometheus/prometheus/retrieval.(*scrapePool).sync(0xc944c52c00, 0xc820e55c18, 0x1, 0x1)
	/go/src/github.com/prometheus/prometheus/retrieval/scrape.go:228 +0x36d
github.com/prometheus/prometheus/retrieval.(*targetSet).sync(0xc9076f29b0)
	/go/src/github.com/prometheus/prometheus/retrieval/targetmanager.go:258 +0x239
github.com/prometheus/prometheus/retrieval.(*targetSet).runProviders(0xc9076f29b0, 0x7ff3f7bb05f0, 0xc94218e0c0, 0xc8ed919530)
	/go/src/github.com/prometheus/prometheus/retrieval/targetmanager.go:334 +0x34b
github.com/prometheus/prometheus/retrieval.(*TargetManager).reload(0xc8202bf3b0)
	/go/src/github.com/prometheus/prometheus/retrieval/targetmanager.go:124 +0x25e
github.com/prometheus/prometheus/retrieval.(*TargetManager).Run(0xc8202bf3b0)
	/go/src/github.com/prometheus/prometheus/retrieval/targetmanager.go:79 +0x1cf
created by main.Main
	/go/src/github.com/prometheus/prometheus/cmd/prometheus/main.go:172 +0x171f
```

--- a/retrieval/targetmanager.go
+++ b/retrieval/targetmanager.go
@@ -331,7 +331,12 @@ func (ts *targetSet) runProviders(ctx context.Context, providers map[string]Targ
 	// We wait for a full initial set of target groups before releasing the mutex
 	// to ensure the initial sync is complete and there are no races with subsequent updates.
 	wg.Wait()
-	ts.sync()
+	// Just signal that there are initial sets to sync now. Actual syncing must only
+	// happen in the runScraping loop.
+	select {
+	case ts.syncCh <- struct{}{}:
+	default:
+	}
 }
 
 // update handles a target group update from a target provider identified by the name.

